extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/PersistenceUnitsHolder.java

```diff
@@ -54,7 +54,13 @@ public static RecordedState popRecordedState(String persistenceUnitName) {
         if (persistenceUnitName == null) {
             key = NO_NAME_TOKEN;
         }
-        return persistenceUnits.recordedStates.remove(key);
+        // FIXME: Changed from remove() to get() to enable DevUI to generate schema scripts.
+        //  The "pop" behavior makes it impossible to construct persistence provider repeatedly (which is needed
+        //  in order to execute `persistenceProvider.generateSchema()`).
+        //  The reason why the recorded state is removed here is to clear away the PU metadata after
+        //  the boot process, when they are no longer needed.
+        //  Would it be OK not to remove the metadata during the dev-mode execution?
+        return persistenceUnits.recordedStates.get(key);
```

Comment:
The "remove" operation is actually important as the memory cost of keeping all metadata around is very large.

Do you think you could try keeping the state only when running in dev mode?

Reply:
Yes, that's what I thought we could do. Yesterday I found out about `LaunchMode.current()` :), so I send another commit that would do `recordedStates.get(key)` or `recordedStates.remove(key)` here depending on current launch mode.